### ISSUE TITLE
query client leak in scale tests

### ISSUE THREAD

**Issue by Laremere** [googleforgames/open-match#1076]
Unclear what's causing it, but it's certainly happening:
https://snapshot.raintank.io/dashboard/snapshot/Ls39viYO7ts8KQJFCDwIRMwdFpaMzrmm?orgId=2

**Comment by Laremere** [comment]
Here's the culprit go routines:
```
goroutine profile: total 32848
10943 @ 0x431290 0x42c02a 0x42b5f5 0x4cc4d5 0x4cd44f 0x4cd431 0x56406f 0x577498 0x5246ba 0x4a96d7 0x8cd777 0x8cd72a 0x8cdfe1 0x8f0d4e 0x9337b3 0x9464af 0x45f021
#	0x42b5f4	internal/poll.runtime_pollWait+0x54						/usr/local/go/src/runtime/netpoll.go:184
#	0x4cc4d4	internal/poll.(*pollDesc).wait+0x44						/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x4cd44e	internal/poll.(*pollDesc).waitRead+0x1ce					/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x4cd430	internal/poll.(*FD).Read+0x1b0							/usr/local/go/src/internal/poll/fd_unix.go:169
#	0x56406e	net.(*netFD).Read+0x4e								/usr/local/go/src/net/fd_unix.go:202
#	0x577497	net.(*conn).Read+0x67								/usr/local/go/src/net/net.go:184
#	0x5246b9	bufio.(*Reader).Read+0x269							/usr/local/go/src/bufio/bufio.go:226
#	0x4a96d6	io.ReadAtLeast+0x86								/usr/local/go/src/io/io.go:310
#	0x8cd776	io.ReadFull+0x86								/usr/local/go/src/io/io.go:329
#	0x8cd729	golang.org/x/net/http2.readFrameHeader+0x39					/go/pkg/mod/golang.org/x/net@v0.0.0-20191105084925-a882066a44e0/http2/frame.go:237
#	0x8cdfe0	golang.org/x/net/http2.(*Framer).ReadFrame+0xa0					/go/pkg/mod/golang.org/x/net@v0.0.0-20191105084925-a882066a44e0/http2/frame.go:492
#	0x8f0d4d	google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams+0x9d	/go/pkg/mod/google.golang.org/grpc@v1.25.0/internal/transport/http2_server.go:451
#	0x9337b2	google.golang.org/grpc.(*Server).serveStreams+0xf2				/go/pkg/mod/google.golang.org/grpc@v1.25.0/server.go:718
#	0x9464ae	google.golang.org/grpc.(*Server).handleRawConn.func1+0x3e			/go/pkg/mod/google.golang.org/grpc@v1.25.0/server.go:679

10943 @ 0x431290 0x440d9b 0x8dc282 0x8dcba3 0x8ff6cb 0x45f021
#	0x8dc281	google.golang.org/grpc/internal/transport.(*controlBuffer).get+0x121	/go/pkg/mod/google.golang.org/grpc@v1.25.0/internal/transport/controlbuf.go:395
#	0x8dcba2	google.golang.org/grpc/internal/transport.(*loopyWriter).run+0x1e2	/go/pkg/mod/google.golang.org/grpc@v1.25.0/internal/transport/controlbuf.go:513
#	0x8ff6ca	google.golang.org/grpc/internal/transport.newHTTP2Server.func2+0xca	/go/pkg/mod/google.golang.org/grpc@v1.25.0/internal/transport/http2_server.go:290

10943 @ 0x431290 0x440d9b 0x8f5066 0x45f021
#	0x8f5065	google.golang.org/grpc/internal/transport.(*http2Server).keepalive+0x1d5	/go/pkg/mod/google.golang.org/grpc@v1.25.0/internal/transport/http2_server.go:954
```
(I've trimmed all other go routines, as they have 2 or fewer count and there's only a handful of them)

Refreshing always results in the numbers going up.

**Comment by Laremere** [comment]
Obtained via pprof:
kubectl port-forward pod/om-query-75f7777f95-tkdxt 51503:51503 -n open-match
http://localhost:51503/debug/pprof/goroutine?debug=1

**Comment by Laremere** [comment]
Something is keeping the connections open.

It's unclear so far whether the clients are keeping them open, or if the clients have gone away, and the connections are just not closing for some reason?
Does this have anything to do with us removing the 60s server timeout?

**Comment by Laremere** [comment]
Unsure if this will solve everything, but my current guess:
0.8 keepalive parameters:
https://github.com/googleforgames/open-match/blob/release-0.8/internal/rpc/server.go#L311
Settings were removed after 0.8:
https://github.com/googleforgames/open-match/blob/master/internal/rpc/server.go#L306
Documentation on keepalive parameters:
https://godoc.org/google.golang.org/grpc/keepalive#ServerParameters

Some of the defaults are:
- clients can be idle for an infinite amount of time.
- clients can be as old as they want to be (this is what used to be 1m, and caused problems.  OM has some potentially very long running rpcs, so infinity seems reasonable here, as long as we're not leaking rpcs).
- Time after no activity to wait for checking that a client is still there: defaults to 2 hours.  <--------- a big problem.

**Comment by Laremere** [comment]
Forgot to save dashboards of running without keepalive, but here's adding a MaxConnectionIdle of 20s:
Scale: https://snapshot.raintank.io/dashboard/snapshot/29uqbCOM8jn27oaYGd2lrIdf2d528ZNC
Go Processes: https://snapshot.raintank.io/dashboard/snapshot/4UxJveTpLMChtwmoUfBn9Inaz5wFE3V6

Still getting into ~50k go routines.  Not sure if that is an improvement or not, but I think it still needs to get better.

**Comment by Laremere** [comment]
At a glance, with 20s idle time, and connections which aren't reused:

1 query /profile * 20 profiles / iteration * 4 iterations / second * 20 seconds before end of keepalive = 1600 queries which haven't failed keepalive.  So that doesn't seemed to have worked as expected.

**Comment by Laremere** [comment]
Scale test mmf currently creates a query client that isn't used: https://github.com/googleforgames/open-match/blob/82a011ea52523a7c1a123e581f39c8483c5bf864/examples/scale/mmf/mmf.go#L42
Then for each mmf run creates another client which is never closed: https://github.com/googleforgames/open-match/blob/1bd63a01c746d7b04dc6e93cdc74f31283266d28/examples/scale/scenarios/util.go#L99

So that's not good.

**Comment by Laremere** [comment]
```
		grpc.KeepaliveParams(keepalive.ServerParameters{
			MaxConnectionIdle: time.Second * 20,
			Time:              time.Second,
			Timeout:           time.Second * 5,
		}),
```
Is exhibiting the same behavior:
Scale test: https://snapshot.raintank.io/dashboard/snapshot/2HtEWZ4QFly8OIdXAwPO5O3Zk2dWYjen
Go processes: https://snapshot.raintank.io/dashboard/snapshot/bOMIk5YI1ruBNDtxFPVfS7gH3n9KyGFW

**Comment by Laremere** [comment]
Yeah, it's the MMF creating and not removing clients (though the keepalive change is good too, to clean up dead connections.)

Using the same query client from the mmf:
Scale test: https://snapshot.raintank.io/dashboard/snapshot/zo47UG3yy2IOzcYNubnBZYxFsciSzFow
Go processes: https://snapshot.raintank.io/dashboard/snapshot/0Pm57x2eoYvxwx9WuMbvLoSm6WCmcoQz

Of note as well is that cpu usage from the matchfunctions drops from ~60 to ~2.

**Comment by yfei1** [comment]
Closed by #1091
